internal/controller/nnf_clientmount_controller.go

```diff
@@ -192,138 +177,80 @@ func (r *NnfClientMountReconciler) changeMountAll(ctx context.Context, clientMou
 		}
 
-		nodeStorage := &nnfv1alpha1.NnfNodeStorage{}
-		if err := r.Get(ctx, namespacedName, nodeStorage); err != nil {
-			return err
+		if clientMount.Spec.Mounts[index].SetPermissions {
+			if err := os.Chown(clientMountInfo.MountPath, int(clientMount.Spec.Mounts[index].UserID), int(clientMount.Spec.Mounts[index].GroupID)); err != nil {
+				return dwsv1alpha2.NewResourceError("unable to set owner and group for file system").WithError(err).WithMajor()
+			}
 		}
-
-		allocationStatus := nodeStorage.Status.Allocations[clientMountInfo.Device.DeviceReference.Data]
-		fileShare, err := r.getFileShare(allocationStatus.FileSystem.ID, allocationStatus.FileShare.ID)
+	} else {
+		unmounted, err := fileSystem.Unmount(ctx, clientMountInfo.MountPath)
 		if err != nil {
-			return dwsv1alpha2.NewResourceError("could not get file share").WithError(err).WithMajor()
+			return dwsv1alpha2.NewResourceError("unable to unmount file system").WithError(err).WithMajor()
 		}
-
-		if shouldMount {
-			fileShare.FileSharePath = clientMountInfo.MountPath
-		} else {
-			fileShare.FileSharePath = ""
+		if unmounted {
+			log.Info("Unmounted file system", "Mount path", clientMountInfo.MountPath)
 		}
-
-		fileShare, err = r.updateFileShare(allocationStatus.FileSystem.ID, fileShare)
-		if err != nil {
-			return dwsv1alpha2.NewResourceError("could not update file share").WithError(err).WithMajor()
-		}
-
-	default:
-		return dwsv1alpha2.NewResourceError("invalid device type %s", clientMountInfo.Device.Type).WithFatal()
-	}
-
-	if shouldMount {
-		log.Info("Mounted file system", "Mount path", clientMountInfo.MountPath)
-	} else {
-		log.Info("Unmounted file system", "Mount path", clientMountInfo.MountPath)
 	}
 
 	return nil
 }
 
-func (r *NnfClientMountReconciler) updateFileShare(fileSystemId string, fileShare *sf.FileShareV120FileShare) (*sf.FileShareV120FileShare, error) {
-	ss := nnf.NewDefaultStorageService()
-
-	if err := ss.StorageServiceIdFileSystemIdExportedShareIdPut(ss.Id(), fileSystemId, fileShare.Id, fileShare); err != nil {
-		return nil, err
+// fakeNnfNodeStorage creates an NnfNodeStorage resource with filled in with only the fields
+// that are necessary to mount the file system. This is done to reduce the API server load
+// because the compute nodes don't need to Get() the actual NnfNodeStorage.
+func (r *NnfClientMountReconciler) fakeNnfNodeStorage(clientMount *dwsv1alpha2.ClientMount, index int) *nnfv1alpha1.NnfNodeStorage {
+	nnfNodeStorage := &nnfv1alpha1.NnfNodeStorage{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clientMount.Spec.Mounts[index].Device.DeviceReference.ObjectReference.Name,
+			Namespace: clientMount.Spec.Mounts[index].Device.DeviceReference.ObjectReference.Namespace,
+		},
 	}
 
-	return fileShare, nil
-}
+	// These labels aren't exactly right (NnfStorage owns NnfNodeStorage), but the
+	// labels that are important for doing the mount are there and correct
+	dwsv1alpha2.InheritParentLabels(nnfNodeStorage, clientMount)
```

Comment:
Should this be fixed?

Reply:
This is done on purpose. It's making a fake NnfNodeStorage resource instead of reading the actual NnfNodeStorage from the API server. If we wanted to read the actual NnfNodeStorage, we'd need the Rabbit's namespace added to the Clientmountd client cache. That would mean the watch()'s in Clientmountd would get events from the Rabbit namespace as well. That would put a lot more load on the API server and increase jitter on the compute. We have enough information contained in the ClientMount resource to populate the important fields in a fake NnfNodeStorage that only lives in memory for the Reconcile() call. A couple of the labels are wrong, but they're not used for anything so it doesn't matter.

Reply:
This fake NnfNodeStorage is not getting the `nnf.cray.hpe.com/pinned_storage_profile_name*` labels.  They aren't on the donor ClientMount that it's using.  So it can't look up the pinned storage profile, to find the mount options it should use.

My workarea is a few commits behind the PR.  I'm using git rev c8260fe.

Reply:
This is what I see for labels on a compute node Clientmount:
```
    dataworkflowservices.github.io/owner.kind: NnfAccess
    dataworkflowservices.github.io/owner.name: global-0-computes
    dataworkflowservices.github.io/owner.namespace: default
    dataworkflowservices.github.io/workflow.name: global
    dataworkflowservices.github.io/workflow.namespace: default
    dataworkflowservices.github.io/workflow.uid: 79783957-f8f1-479f-a056-b160661a95aa
    nnf.cray.hpe.com/directive_index: "0"
    nnf.cray.hpe.com/pinned_storage_profile_name: global-0
    nnf.cray.hpe.com/pinned_storage_profile_namespace: default
    nnf.cray.hpe.com/target_directive_index: "0"
```

That hasn't changed for a while, so I'm not sure why you're seeing that. What's your #DW line?

Reply:
My #DW lines look like this:
```
    - "#DW jobdw type=xfs capacity=1GB name=stg1"
    - "#DW copy_out source=$DW_JOB_stg1 destination=/lus/foo"
```
But I'm also running the clientmountd on my mac...so I guess I still don't have that working correctly.